File: lambda-without-type/js/trees.js
'use strict';

export class AST {
  constructor(info) {
    this.info = info;
    this.pos = info.pos;
    this.highlighted = info.highlighted;
  }
}

export class VarTree extends AST {
  constructor(info, n) {
    super(info);
    this.name = n;
    Object.freeze(this);
  }

  toString() { return this.name; }
  toTreeString() { return `Var(${this.name})`; }
}

export class AbsTree extends AST {
  constructor(info, arg, body) {
    super(info);
    this.arg = arg;
    this.body = body;
    Object.freeze(this);
  }

  toString({ isLast = true } = {}) {
    let s = `\\${this.arg}.${this.body.toString({ isFirst: true, isLast: true })}`;
    if(!isLast) s = `(${s})`;
    return s;
  }

  toTreeString() { return `Abs(${this.arg}, ${this.body.toTreeString()})`; }
}

export class AppTree extends AST {
  constructor(info, fn, arg) {
    super(info);
    this.fn = fn;
    this.arg = arg;
    Object.freeze(this);
  }

  toString({ isFirst = true, isLast = true } = {}) {
    const fnStr = this.fn.toString({ isFirst: true, isLast: false });
    const argStr = this.arg.toString({ isFirst: false, isLast: isLast || !isFirst });

    let s = `${fnStr} ${argStr}`;
    if(!isFirst) s = `(${s})`;
    return s;
  }

  toTreeString() { return `App(${this.fn.toTreeString()}, ${this.arg.toTreeString()})`; }
}

File: lambda-without-type/hs/trees.hs
module Lam.Trees (AST(..), showLam) where

data AST = VarTree String
         | AbsTree String AST
         | AppTree AST AST deriving Show

showLam :: AST -> String
showLam = showLam' True True
  where
    -- showLam' isFirstChild isLastChild tree
    showLam' :: Bool -> Bool -> AST -> String
    showLam' _ _ (VarTree v) = v
    showLam' _ True (AbsTree arg body) = "\\" ++ arg ++ "." ++ showLam' True True body
    showLam' True l (AppTree fn arg) = showLam' True False fn ++ " " ++ showLam' False l arg
    showLam' _ _ t = "(" ++ showLam' True True t ++ ")"
